src/Microsoft.Identity.Web/ClientAssertionDescription.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Description of a client assertion in the application configuration.
+    /// See https://aka.ms/ms-id-web/client-assertions.
+    /// </summary>
+    public class ClientAssertionDescription
+    {
+        /// <summary>
+        /// Constructor of a ClientAssertionDescription.
+        /// </summary>
+        /// <param name="clientAssertionProvider">delegate providing the client assertion
+        /// when it is necessary.</param>
+        public ClientAssertionDescription(Func<CancellationToken, Task<ClientAssertion>> clientAssertionProvider)
```

Comment:
Having played around with client assertion recently, I think this "func" mechanism is not very good. I am thinking of making the following changes in MSAL: 

- all customers to specify client assertion on the request (i.e. AcquireToken*). 
- move away from FUNC to interface (this is semantically equivalent but interfaces are easier to reason with)
- interface will have a single method with 2 input params instead of 1:

`Task<string> GetClientAssertionAsync(string audience, CancellationToken token)`

This is because computing the audience (which is the token endpoint string) is quite complicated, but MSAL already knows it at the moment it needs the assertion. 

Thoughts?